app/workers/work_packages/update_progress_job.rb

```diff
@@ -0,0 +1,257 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+# rubocop:disable Rails/SquishedSQLHeredocs
+class WorkPackages::UpdateProgressJob < ApplicationJob
+  queue_with_priority :default
+
+  def perform(current_mode:, previous_mode:)
+    with_temporary_progress_table do
+      if current_mode == "field"
+        unset_all_percent_complete_values if previous_mode == "disabled"
+        fix_remaining_work_set_with_100p_complete
+        fix_remaining_work_exceeding_work
+        fix_only_work_being_set
+        fix_only_remaining_work_being_set
+        derive_unset_remaining_work_from_work_and_p_complete
+        derive_unset_work_from_remaining_work_and_p_complete
+        derive_p_complete_from_work_and_remaining_work
+      end
+      if current_mode == "status"
+        set_p_complete_from_status
+        fix_remaining_work_set_with_100p_complete
+        derive_unset_work_from_remaining_work_and_p_complete
+        derive_remaining_work_from_work_and_p_complete
+      end
+
+      update_totals
+
+      copy_progress_values_to_work_packages_and_update_journals
+    end
+  end
+
+  private
+
+  def with_temporary_progress_table
+    create_temporary_progress_table
+    yield
+  ensure
+    drop_temporary_progress_table
+  end
+
+  def create_temporary_progress_table
+    execute(<<~SQL)
+      CREATE UNLOGGED TABLE temp_wp_progress_values
+      AS SELECT
+        id,
+        status_id,
+        estimated_hours,
+        remaining_hours,
+        done_ratio,
+        NULL::double precision AS total_work,
+        NULL::double precision AS total_remaining_work,
+        NULL::integer AS total_p_complete
```

Comment:
I'd include the `updated_at` or `lock_version` here. Otherwise, a user might loose their input in case they update the work package between the time the `temp_...` table is filled and then written back into the `work_packages` table. When writing back, the sql could bail on rows that have a higher `updated_at`/`lock_version` than the one in the `temp_...` table.

Reply:
Indeed, this is a possibility.

It should then be made more robust.
I'll add it on the TODO.